### PR TITLE
executor: Reduce the memory usage of InsertExec

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 

InsertExec consumes too much memory [14355](https://github.com/pingcap/tidb/issues/14355)

### What is changed and how it works?

allocate less rows and batch exec.
we can see memory usage of insert 500,000 rows reduce from 2.3GB to 1.7GB
image

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

### CODE REVIEW

**Comment by shenli** [comment]
@b41sh Thanks for your amazing work!

**Comment by jackysp** [comment]
Could you show the difference of go pprof heap?  @b41sh

**Comment by b41sh** [comment]
image
image

alloc_space of insertRows drop from 11.45MB to 0.77MB
@jackysp PTAL

**Comment by jackysp** [comment]
Thank, @b41sh ! But it doesn't seem to be a great improvement. Can you analyze the reason? Because this modification may bring some negative effects, such as making the insert ignore / on duplicate update statement call the BatchGet interface more times. We needs to better balance pros and cons.

File: executor/insert_common.go [hunk 1]
```diff
@@ -239,13 +242,15 @@ func insertRows(ctx context.Context, base insertCommon) (err error) {
 			rows = rows[:0]
 			memTracker.Consume(-memUsageOfRows)
 			memUsageOfRows = 0
-			if err = e.doBatchInsert(ctx); err != nil {
-				return err
+			if batchInsert {
+				if err = e.doBatchInsert(ctx); err != nil {
+					return err
+				}
```
Comment (qw4990) [inline executor/insert_common.go]: Why do we change the behavior of `batchInsert`?

Comment (qw4990) [reply executor/insert_common.go]: Is it better to keep the same behavior we use before?

**Comment by b41sh** [comment]
@jackysp @XuHuaiyu @qw4990 thanks for your guidance.
As the main use of memory is in `e.Lists`, and `rows` reset does not immediately reclaim memory. This PR will not bring improvement, so I close it.

**Comment by XuHuaiyu** [comment]
- It's my mistake here. I thought the most of the memory space is consumed during the execution phase, but actually:
  - 79.56%: parse phase and planBuilder phases
  - 8.34%: transaction buffer, and mutation during 2pc
屏幕快照 2020-02-12 下午5 42 07

- I bench the heap of branch master and this commit, this commit does not bring an actual reduction of the inuse_space. 
branch master:
```
(pprof) list execute
Total: 771.81MB
ROUTINE ======================== github.com/pingcap/tidb/session.(*session).execute in /home/tidb/xhy/GOPATH/src/github.com/pingcap/tidb/session/session.go
         0   659.09MB (flat, cum) 85.39% of Total
```

this commit:
```
(pprof) list execute
Total: 938.31MB
ROUTINE ======================== github.com/pingcap/tidb/session.(*session).execute in /home/tidb/xhy/GOPATH/src/github.com/pingcap/tidb/session/session.go
         0   825.72MB (flat, cum) 88.00% of Total
```

**Pull request closed without merging**